waveorder/util.py

```diff
@@ -871,52 +872,180 @@ def Dual_variable_ADMM_TV_deconv_2D(AHA, b_vec, rho, lambda_u, lambda_p, itr, ve
+                           
+        search_range_auto: float
+                           (if using autotune) L-curve search will occur on the range of reg_re +/- search_range_auto
+                           
+        verbose          : bool
+                           option to display detailed progress of computations or not
     
     Returns
     -------
-        f_real   : numpy.ndarray
-                   3D unscaled phase reconstruction with the size of (Ny, Nx, Nz)
+        f_real           : numpy.ndarray
+                           3D unscaled phase reconstruction with the size of (Ny, Nx, Nz)
+                           (if using autotune) reconstruction for the automatically chosen parameter, plus two others around that parameter value, size (3, Ny, Nx, Nz)
     '''
-    
-    if use_gpu:
-        
+    if use_gpu:     
         globals()['cp'] = __import__("cupy")
         cp.cuda.Device(gpu_id).use()
-        
-        S0_stack_f = cp.fft.fftn(cp.array(S0_stack.astype('float32')), axes=(-3,-2,-1))
-        H_eff      = cp.array(H_eff.astype('complex64'))
-        
-        f_real     = cp.asnumpy(cp.real(cp.fft.ifftn(S0_stack_f * cp.conj(H_eff) / (cp.abs(H_eff)**2 + reg_re),axes=(-3,-2,-1))))
+        S0_stack = cp.array(S0_stack.astype('float32'))
+        H_eff = cp.array(H_eff.astype('complex64'))
+        xp = cp
     else:
+        xp = np
+    
+    S0_stack_f = xp.fft.fftn(S0_stack, axes=(-3,-2,-1))
+    N,M,L = S0_stack_f.shape
+    
+    # a "named tuple" representing a point on the L curve
+        # allows for dot notation to access attributes
+    Point_L_curve = namedtuple("Point_L_curve", "reg_x reg data_norm reg_norm f_real_f")
+    
+    def menger_curvature(points):
+        # x is data norm, y is reg norm
+        x, y = xp.zeros(3), xp.zeros(3)
+        for i in range(3):
+            x[i] = points[i].data_norm
+            y[i] = points[i].reg_norm
+        d0 = ((x[1]-x[0])**2+(y[1]-y[0])**2)**(1/2)
+        d1 = ((x[2]-x[1])**2+(y[2]-y[1])**2)**(1/2)
+        d2 = ((x[2]-x[0])**2+(y[2]-y[0])**2)**(1/2)
+
+        signed_area = 0.5*( (x[1]-x[2])*(y[1]-y[0])-(x[1]-x[0])*(y[1]-y[2]) )
+
+        return 4*signed_area / (d0*d1*d2)
+    
+    # evaluate the L curve at a specific lambda
+        # returns a new Point_L_curve() tuple.
+        # this function is the only place where new points are instantiated
+    def eval_L_curve(reg_x):
+        reg_coeff = 10**reg_x
+        H_eff_abs_square = xp.abs(H_eff)**2
+        H_eff_conj = xp.conj(H_eff)
+
+        # FT{f} (f=scattering potential (whose real part is (scaled) phase))
+        f_real_f = S0_stack_f * H_eff_conj / (H_eff_abs_square + reg_coeff)
+        S0_est_stack_f = H_eff * f_real_f # Ax (put estimate through forward model)
+
+        data_norm_eval = xp.log(xp.linalg.norm(S0_est_stack_f - S0_stack_f)**2 /N/M/L)
+        reg_norm_eval = xp.log(xp.linalg.norm(f_real_f)**2 /N/M/L)
+
```

Comment:
Line 956--960 are only used for evaluating points on L-curve but are not used for reconstruction without auto-tuning. Maybe consider creating a function specifically for reconstruction and the other function for L-curve evaluation that will call the reconstruction function when calculating `f_real_f`?